examples/dpo_demo_gemma3.ipynb

```diff
@@ -0,0 +1,849 @@
+    "            chosen_ids, chosen_mask = _generate_ids_and_masks(\n",
+    "                [example[\"chosen\"]], tokenizer, total_generation_steps, left_pad=False\n",
+    "            )\n",
+    "            rejected_ids, rejected_mask = _generate_ids_and_masks(\n",
+    "                [example[\"rejected\"]], tokenizer, total_generation_steps, left_pad=False\n",
+    "            )\n",
+    "            processed_examples.append({\n",
+    "                \"prompt_ids\": prompt_ids.astype(np.int64),\n",
+    "                \"prompt_mask\": prompt_mask.astype(np.float64), # Cast to float64\n",
+    "                \"chosen_ids\": chosen_ids.astype(np.int64),\n",
+    "                \"chosen_mask\": chosen_mask.astype(np.float64), # Cast to float64\n",
+    "                \"rejected_ids\": rejected_ids.astype(np.int64),\n",
+    "                \"rejected_mask\": rejected_mask.astype(np.float64), # Cast to float64\n",
+    "            })\n",
+    "\n",
+    "        data_dict = {key: np.array([example[key] for example in processed_examples]) for key in processed_examples[0].keys()}\n",
+    "\n",
+    "\n",
+    "        training_input = TrainingInput(\n",
+    "            prompt_ids=data_dict[\"prompt_ids\"],\n",
+    "            prompt_mask=data_dict[\"prompt_mask\"],\n",
+    "            chosen_ids=data_dict[\"chosen_ids\"],\n",
+    "            chosen_mask=data_dict[\"chosen_mask\"],\n",
+    "            rejected_ids=data_dict[\"rejected_ids\"],\n",
+    "            rejected_mask=data_dict[\"rejected_mask\"],\n",
+    "        )\n",
+    "        processed_batches.append(training_input)\n",
+    "\n",
+    "    return processed_batches\n",
+    "\n",
+    "processed_dpo_dataset = process_dpo_dataset(combined_dpo_dataset, gemma_tokenizer, MAX_PROMPT_LENGTH, TOTAL_GENERATION_STEPS, BATCH_SIZE)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class MySource():\n",
+    "\n",
+    "  def __init__(self, data):\n",
+    "    self._data = data\n",
+    "\n",
+    "  def __getitem__(self, idx):\n",
+    "    return self._data[idx]\n",
+    "\n",
+    "  def __len__(self):\n",
+    "    return len(self._data)\n",
+    "def _dummy_dataset(\n",
+    "    source: MySource,\n",
+    "    prompt_ids: np.ndarray,\n",
+    "    prompt_mask: np.ndarray,\n",
+    "    chosen_ids: np.ndarray,\n",
+    "    chosen_mask: np.ndarray,\n",
+    "    rejected_ids: np.ndarray,\n",
+    "    rejected_mask: np.ndarray,\n",
+    "):\n",
+    "  return grain.MapDataset.source(source).map(\n",
+    "      lambda x: TrainingInput(\n",
+    "          prompt_ids=prompt_ids[x],\n",
+    "          prompt_mask=prompt_mask[x],\n",
+    "          chosen_ids=chosen_ids[x],\n",
+    "          chosen_mask=chosen_mask[x],\n",
+    "          rejected_ids=rejected_ids[x],\n",
+    "          rejected_mask=rejected_mask[x],\n",
+    "      )\n",
```

Comment:
I believe a lot of the preprocessing code can be simplified here. Refer to this: https://github.com/google/tunix/blob/main/examples/dpo_demo.ipynb